lib/src/widgets/board_editor.dart

```diff
@@ -0,0 +1,232 @@
+import 'package:chessground/src/board_geometry.dart';
+import 'package:flutter/widgets.dart';
+
+import 'piece.dart';
+import 'positioned_square.dart';
+import '../models.dart';
+import '../fen.dart';
+import '../board_editor_settings.dart';
+
+/// A chessboard widget whose pieces can be dragged around freely.
+///
+/// This widget can be used as the basis for a fully fledged board editor, similar to https://lichess.org/editor.
+class BoardEditor extends StatefulWidget with BoardGeometry {
+  const BoardEditor({
+    super.key,
+    required this.size,
+    required this.orientation,
+    required this.initialFen,
+    this.pieceToAddOnTap,
+    this.removePieceOnTap = false,
+    this.settings = const BoardEditorSettings(),
+    this.onPiecesChanged,
+  });
+
+  /// Visual size of the board.
+  @override
+  final double size;
+
+  /// Settings that control the appearance of the board editor.
+  final BoardEditorSettings settings;
+
+  /// Side by which the board is oriented.
+  @override
+  final Side orientation;
+
+  /// FEN string describing the initial position of the board editor.
+  final String initialFen;
+
+  /// If true, pieces will be removed when tapped. This takes precedence over [pieceToAddOnTap].
+  final bool removePieceOnTap;
+
+  /// If not null, this piece will be added when a square is tapped, potentially replacing an existing piece.
+  final Piece? pieceToAddOnTap;
+
+  /// Callback to get notified when the position changes, by a piece being moved, discarded or added.
+  ///
+  /// A full board editor implementation can use this to keep track of the current position and export it as a FEN string if needed.
+  final void Function(Pieces)? onPiecesChanged;
+
+  @override
+  // ignore: library_private_types_in_public_api
+  _BoardEditorState createState() => _BoardEditorState();
+}
+
+class _BoardEditorState extends State<BoardEditor> {
+  Pieces pieces = {};
+
+  /// Avatar for the piece that is currently being dragged.
+  DragAvatar? _dragAvatar;
+
+  /// Once a piece is dragged, holds the square id of the piece.
+  SquareId? _draggedPieceSquareId;
+
+  /// Current render box during drag.
+  RenderBox? _renderBox;
+
+  @override
+  Widget build(BuildContext context) {
+    final List<Widget> pieceWidgets = [
+      for (final MapEntry(key: squareId, value: piece) in pieces.entries)
+        if (squareId != _draggedPieceSquareId)
+          PositionedSquare(
+            key: ValueKey('$squareId-${piece.kind.name}'),
+            size: widget.squareSize,
+            orientation: widget.orientation,
+            squareId: squareId,
+            child: PieceWidget(
+              piece: piece,
+              size: widget.squareSize,
+              pieceAssets: widget.settings.pieceAssets,
+            ),
+          ),
+    ];
+
+    widget.onPiecesChanged?.call(pieces);
```

Comment:
This is probably not the most idiomatic way to do this. Also, instead of sending all the pieces, should we go for delta updates instead? (add/remove) This would be a bit more work for the library consumer though, and would also have higher chances of bugs being introducing by things getting out of sync

Reply:
Indeed you should not perform side effects in `build()`. But I think my api design suggestion in the other comment answers that concern.